lib/lotus/logger.rb

```diff
@@ -0,0 +1,88 @@
+require 'logger'
+
+module Lotus
+  # Lotus default logger
+  #
+  # Implement with the same interface of ruby std ::Logger
+  # Optout the logdev parameter and always use STDOUT to output the log
+  #
+  # Lotus logger also has the app_tag concept, which used to identify
+  # which application the log come from
+  #
+  # Example Usage:
+  # #1 Logger with namespace
+  # module App
+  #   class AppLogger < Lotus::Logger
+  #   end
+  #
+  #   def foo
+  #     Applogger.new.info('foo')
+  #     #=> output: I, [2015-01-10T21:55:12.727259 #80487]  INFO -- [App] : foo
+  #   end
+  # end
+  #
+  # #2 Logger without namespace
+  # class AppLogger < Lotus::Logger
+  # end
+  #
+  # Applogger.new.info('foo')
+  # #=> output: I, [2015-01-10T21:55:12.727259 #80487]  INFO -- [Shared] : foo
+  #
+  # @since 0.1.0
+
+  class Logger < ::Logger
+
+    # Override Ruby's Logger#initialize
+    #
+    # @param logdev is default to STDOUT
+    #
+    # @since 0.2.1
+    def initialize(logdev=nil, *args)
+      super($stdout, *args)
+      @formatter = LotusFormatter.new.tap {|f| f.app_tag = self.app_tag}
+    end
+
+    # app_tag is the identification of current app
+    # app_tag is default to use highest namespace if current namespace
+    # if app_tag is blank, lotus use default app_tag 'shared'
+    # @param logdev is default to STDOUT
+    #
+    # @since 0.2.1
+    def app_tag
+      _app_tag
+    end
+
+    # Lotus Logger default Formatter
+    #
+    # inject app_tag to logged messages
+    #
+    # @since 0.2.1
+    class LotusFormatter < ::Logger::Formatter
+      attr_accessor :app_tag
+
+      def call(severity, time, progname, msg)
+        raise Exception.new('You must set the app_tag') unless (app_tag && !app_tag.empty?)
```

Comment:
`Exception` should never be used.

The reason is that anonymous `rescue` clause won't catch this kind of exceptions.

```ruby
[1] pry(main)> begin
[1] pry(main)*   raise Exception.new('boom')
[1] pry(main)* rescue
[1] pry(main)*   puts 'rescued'
[1] pry(main)* end
Exception: boom
from (pry):2:in `__pry__'
[2] pry(main)>
```

If we don't care to have a custom exception for this case (and I agree), we should use `raise "You must set the app_tag"`. This syntax raises a `RuntimeError`, which is catchable by anonymous `rescue`.

```ruby
[1] pry(main)> begin
[1] pry(main)*   raise "missing app_name"
[1] pry(main)* rescue => e
[1] pry(main)*   puts e.class.name
[1] pry(main)*   puts e.message
[1] pry(main)* end
RuntimeError
missing app_name
=> nil
[2] pry(main)>
```

This pattern should always be used by libraries and frameworks, so developers should never worry to use specific class names with `rescue`.

More at:

* http://robots.thoughtbot.com/rescue-standarderror-not-exception
* http://daniel.fone.net.nz/blog/2013/05/28/why-you-should-never-rescue-exception-in-ruby/
* http://www.mikeperham.com/2012/03/03/the-perils-of-rescue-exception/